server/routes/v1/tokenDiscovery.js

```diff
@@ -114,21 +125,63 @@ async function cacheTokensList() {
   }
 }
 
-if (!timerStarted && process.env.REACT_APP_ENVIRONMENT === 'mainnet') {
-  timerStarted = true;
-  cacheTokensList();
-  setInterval(() => cacheTokensList(), TIME_INTERVAL);
+// starts the caching process for bigquery
+function startCaching() {
+  // only run if on mainnet (the tokens page doesn't exist on devnet/testnet)
+  if (process.env.REACT_APP_ENVIRONMENT === 'mainnet') {
+    // initialize the cache
+    cacheTokensList();
+    // cache every TIME_INTERVAL ms (only starts after one interval)
+    const intervalId = setInterval(() => cacheTokensList(), TIME_INTERVAL);
+    // Stop the auto-running of the caching in the previous line after TIME_TO_TEST ms
+    // Only do this if not in the prod env, so we don't have excessive BigQuery queries when we're
+    // not actually using what's in the dev and staging environments
+    // We don't want regular caching to stop in prod, though, because then a missed cache would
+    // result in a several-second delay while the query is re-run, and this is a poor UX
+    if (!IS_PROD_ENV) {
+      setTimeout(() => {
+        log.info('stopping caching tokens');
+        clearInterval(intervalId);
+      }, TIME_TO_TEST);
+    }
+  }
 }
 
+startCaching();
+
 function sleep(ms) {
   return new Promise(resolve => setTimeout(resolve, ms));
 }
 
 module.exports = async (req, res) => {
+  // * * * * * * * * IMPORTANT * * * * * * * * * * * *
+  // Running the BigQuery query costs money. Don't let it run if you don't need to.
   log.info(`getting token discovery`);
   try {
+    // if it's been a while since caching happened in the non-prod envs, then restart the caching
+    // (needed because `startCaching` turns off caching after TIME_TO_TEST for non-prod envs)
+    if (
+      // true if in dev or staging
+      !IS_PROD_ENV &&
+      // true if the cache has already been initialized (i.e. if `startCaching` has already run at
+      // least once)
+      // (prevents race conditions where an API call is made on container startup before the
+      // initial `startCaching` has returned)
+      cachedTokensList.time != null &&
+      // true if some time has passed since the cache was last updated (i.e. `startCaching` was
+      // turned off)
+      // this uses `TIME_INTERVAL * 2` to prevent race conditions around `TIME_INTERVAL`, in case
+      // the cache just took some time to load but `startCaching` has already been triggered
+      Date.now() - cachedTokensList.time > TIME_INTERVAL * 2
```

Comment:
In-lining comments in the if-statement makes it kinda hard to read. It might be better to pull out the relevant terms into variables with comments, then just combine them in the if-statement.

Reply:
i'm with jackson